### ISSUE TITLE
Analyze parser performance degradation with recent Vim versions

### ISSUE THREAD

**Issue by bohlender** [bohlender/vim-z3-smt2#8]
This issue is restricted to the (optional) formatting functionality of the plugin.

- The first step of auto-formatting an `*.smt2` file is parsing it
- For reference, parsing a file of 27KB used to take **<0.01s** with Vim **v8.2.2725**.
- Newer versions seem to be a bit slower, i.e. Vim **v9.0.1107** needs **0.5s** to parse the same file.
- With Vim **v9.0.1122**, parsing the same file suddenly takes **17.5s** :warning:
- The latest release Vim **v9.1.0912** is even slightly worse: **19s**

The plugin's code didn't change -- it's Vim getting worse at executing vim9script.
We've got to at least remedy whatever happened between v9.0.1107 and v9.0.1122.

The **goal of this task** is to:
- provide a minimal example, s.t. the Vim folks can investigate the performance degradation, or
- pinpoint the problematic commit in the [Vim repo](https://github.com/vim/vim)

_Note to myself: Tested with `docker run --rm -it -v ~/:/root thinca/vim:v9.0.1107 /root/Desktop/smt2/append_fs_unsafe.c.mini.smt2`_

**Comment by bohlender** [comment]
According to [`:help patches-9.1`](https://vimhelp.org/version9.txt.html#patches-9) one of these changes must have introduced the performance degradation:
```
Patch 9.0.1108
Problem:    Type error when using "any" type and adding [a](https://vimhelp.org/insert.txt.html#a) number to [a](https://vimhelp.org/insert.txt.html#a) float.
Solution:   Accept both [a](https://vimhelp.org/insert.txt.html#a) number and [a](https://vimhelp.org/insert.txt.html#a) float. (closes #11753)

Patch 9.0.1109
Problem:    Leaking allocated type.
Solution:   Reset the "static" flag in the allocated type copy.

Patch 9.0.1110
Problem:    Build fails on [Mac](https://vimhelp.org/os_mac.txt.html#Mac) OS [X](https://vimhelp.org/change.txt.html#X) 10.4/10.5 [.](https://vimhelp.org/repeat.txt.html#.)
Solution:   Check if the dispatch/dispatch.h header exists. (Evan Miller,
            closes #11746)

Patch 9.0.1111
Problem:    Termcap entries for RGB colors are not set automatically.
Solution:   Always set the [termcap](https://vimhelp.org/term.txt.html#termcap) entries when [+termguicolors](https://vimhelp.org/various.txt.html#%2Btermguicolors) [is](https://vimhelp.org/motion.txt.html#is) enabled.

Patch 9.0.1112
Problem:    [test_mswin_event()](https://vimhelp.org/testing.txt.html#test_mswin_event%28%29) can hang.
Solution:   Add the "execute" argument to process events right away.
            (Christopher Plewright, closes #11760)

Patch 9.0.1113
Problem:    Users cannot e[a](https://vimhelp.org/insert.txt.html#a)sily try out a PR.
Solution:   Add an "artifacts" [section](https://vimhelp.org/motion.txt.html#section) to the AppVeyor CI config. (Christian
            Brabandt, closes #11762)

Patch 9.0.1114
Problem:    CI does not use the latest [Python](https://vimhelp.org/if_pyth.txt.html#Python) version.
Solution:   Switch from [Python](https://vimhelp.org/if_pyth.txt.html#Python) 3.10 to 3.11. (closes #11761)

Patch 9.0.1115
Problem:    Code [is](https://vimhelp.org/motion.txt.html#is) indented more than needed.
Solution:   Use an early return to reduce indenting. (Yegappan Lakshmanan,
            closes #11758)

Patch 9.0.1116
Problem:    Compiler may complain about an unused function.
Solution:   Add #ifdef. (John Marriott)

Patch 9.0.1117
Problem:    Terminfo entries for bracketed paste are not used.
Solution:   Use the newly added [terminfo](https://vimhelp.org/term.txt.html#terminfo) entries for bracketed paste.
            Correct mixup of output strings and key codes.

Patch 9.0.1118
Problem:    Sporadic test failures when using [a](https://vimhelp.org/insert.txt.html#a) [terminal](https://vimhelp.org/terminal.txt.html#terminal) window.
Solution:   Adjust waiting times. (James McCoy, closes #11763)

Patch 9.0.1119
Problem:    Type of arguments not checked when c[a](https://vimhelp.org/insert.txt.html#a)lling a partial.
Solution:   Give an error for [a](https://vimhelp.org/insert.txt.html#a) wrong argument type. (closes #11753)

Patch 9.0.1120
Problem:    Tex [filetype](https://vimhelp.org/filetype.txt.html#filetype) detection not sufficiently tested.
Solution:   Add more test cases for "tex" detection. (Jonas Strittmatter,
            closes #11765)

Patch 9.0.1121
Problem:    Cursor positioning and display problems with ['smoothscroll'](https://vimhelp.org/options.txt.html#%27smoothscroll%27) and
            using "[zt](https://vimhelp.org/scroll.txt.html#zt)"[,](https://vimhelp.org/motion.txt.html#%2C) "[zb](https://vimhelp.org/scroll.txt.html#zb)" or "[zz](https://vimhelp.org/scroll.txt.html#zz)"[.](https://vimhelp.org/repeat.txt.html#.)
Solution:   Adjust computations and conditions. (Yee Cheng Chin,
            closes #11764)

Patch 9.0.1122
Problem:    [Class](https://vimhelp.org/vim9class.txt.html#Class) member access [is](https://vimhelp.org/motion.txt.html#is) not fully tested yet.
Solution:   Add more tests.
```

**Comment by bohlender** [comment]
The problematic transition seems to be from Vim v9.0.1118 to v9.0.1119:
[![asciicast](https://asciinema.org/a/693844.svg)](https://asciinema.org/a/693844)

According to [`:help patches-9.1`](https://vimhelp.org/version9.txt.html#patches-9) this is the change:
```
Patch 9.0.1119
Problem:    Type of arguments not checked when c[a](https://vimhelp.org/insert.txt.html#a)lling a partial.
Solution:   Give an error for [a](https://vimhelp.org/insert.txt.html#a) wrong argument type. (closes #11753)
```

The linked issue actually even mentions that [the patch "adds some overhead"](https://github.com/vim/vim/issues/11753#issuecomment-1368219440) :see_no_evil:
